File: RPS/source/rps_pose.py
#!/usr/bin/env python

import rospy
import cv2
import numpy
import yaml

from geometry_msgs.msg import PoseStamped
from sensor_msgs.msg import Image
from cv_bridge import CvBridge

# calculation of camera's position, orientation and velocity

def rps_get_pose_msg(self,frame_id):
	
	msg = PoseStamped()
	msg.header.stamp = rospy.get_rostime()
	msg.header.frame_id = frame_id
	
	if frame_id != '':
		msg.pose.position.x = self.pose[0]
		msg.pose.position.y = self.pose[1]
		msg.pose.position.z = self.pose[2]

		msg.pose.orientation.x = self.orient[0]
		msg.pose.orientation.y = self.orient[1]
		msg.pose.orientation.z = self.orient[2]
		msg.pose.orientation.w = 1

	return msg


def rps_calculate_and_publish(self):

	# calculate pose and velocity
	if self.markers_detected > 0:
		self.fail_cnt = 0
		self.pose = -self.tvec
		self.orient = self.rvec

	else:
		if not (self.fail_cnt is None):
			self.fail_cnt += 1

			if self.fail_cnt == 5:
				self.pose = None
				self.fail_cnt = None

		else:
			self.pose = None

	# publish pose
	if not(self.pose is None):
		msg = self.get_pose_msg('main_camera')
	else:
		msg = self.get_pose_msg('') 
	
	self.pose_pub.publish(msg)

#--- detection of aruco markers ---#

def rps_detect(self,data):

	# detect
	self.raw_img = self.bridge.imgmsg_to_cv2(data, 'bgr8')
	corners, ids, rejected_objects = cv2.aruco.detectMarkers(self.raw_img, self.dict, parameters=self.param)

	# calculate rvec, tvec -> pose, orient, vel
	if not (ids is None):
		if not (self.rvec is None) and not(self.tvec is None):
			self.markers_detected, self.rvec, self.tvec = cv2.aruco.estimatePoseBoard(corners, ids, self.board, self.camera_matrix,self.dist_coeffs, self.rvec, self.tvec, useExtrinsicGuess=True) 
		else:
			self.markers_detected, self.rvec, self.tvec = cv2.aruco.estimatePoseBoard(corners, ids, self.board, self.camera_matrix,self.dist_coeffs)
	else:
		self.markers_detected = 0

	self.calculate_and_publish()

	# debug
	#if self.markers_detected > 0:
		#print('number = {0}, x = {1:.3f}, y = {2:.3f}, z = {3:.3f}'.format(self.markers_detected, -self.tvec[0][0], -self.tvec[1][0], -self.tvec[2][0]))
	#else:
		#print('no pose')

	# send img to /RPS/debug
	if self.img_pub.get_num_connections() > 0:
		self.img = cv2.aruco.drawDetectedMarkers(self.raw_img, corners, ids)
		if self.markers_detected > 0:
			self.img = cv2.aruco.drawAxis(self.img, self.camera_matrix, self.dist_coeffs, self.rvec, self.tvec, 0.4)
		self.img_pub.publish(self.bridge.cv2_to_imgmsg(self.img, 'bgr8'))

	# image of board -> /RPS/board_image
	if not(self.board_img is None) and (self.board_img_pub.get_num_connections() > 0):
		self.board_img_pub.publish(self.bridge.cv2_to_imgmsg(self.board_img, 'mono8'))

#--- board and camera matrix ---#

def rps_calculate_board(self):

	# parameters of board
	number_x = int(rospy.get_param('~markers_number_x', '1'))
	number_y = int(rospy.get_param('~markers_number_y', '1'))
	first = int(rospy.get_param('~first_marker', '0'))
	ids = [[i] for i in range(first,first + number_x*number_y)]
	side = float(rospy.get_param('~markers_side', '0.22'))
	dist_x = float(rospy.get_param('~markers_dist_x', '0'))
	dist_y = float(rospy.get_param('~markers_dist_y', '0'))

	# actions
	obj_points = []
	corners = [[],[],[],[]]
	max_y = side + (number_y - 1)*(side + dist_y)
	for i in range(number_y):
		for j in range(number_x):
			corners[0] = [j*(side + dist_x), max_y - i*(side + dist_y), 0]
			corners[1] = [j*(side + dist_x) + side, max_y - i*(side + dist_y), 0]
			corners[2] = [j*(side + dist_x) + side, max_y - i*(side + dist_y) - side, 0]
			corners[3] = [j*(side + dist_x), max_y - i*(side + dist_y) - side, 0]
			corners = numpy.asarray(corners, dtype=numpy.float32)
			obj_points.append(corners.copy())

	ids = numpy.asarray(ids)

	return obj_points,ids

def rps_create_board(self):

	# create board
	obj_points,ids = self.calculate_board()
	self.board = cv2.aruco.Board_create(obj_points, self.dict, ids)

	# create image of board
	self.board_img = cv2.aruco.drawPlanarBoard(self.board, (1000,1000))

def rps_correct_cm(self):
	vector = self.camera_matrix['data']
	matrix = [[],[],[]]
	for i in range(3):
		for j in range(3):
			matrix[i].append(vector[3*i + j])
	self.camera_matrix = numpy.asarray(matrix, dtype=numpy.float64)

#--- main class ---#

class rps_class:

	detect = rps_detect
	create_board = rps_create_board
	calculate_board = rps_calculate_board
	correct_cm = rps_correct_cm

	get_pose_msg = rps_get_pose_msg
	calculate_and_publish = rps_calculate_and_publish

	def __init__(self):

		rospy.loginfo('initializing rps_pose...')

		self.dict = cv2.aruco.Dictionary_get(cv2.aruco.DICT_4X4_1000)
		self.param = cv2.aruco.DetectorParameters_create()
		self.board = None

		self.bridge = CvBridge()
		self.raw_img = None
		self.img = None
		self.board_img = None

		self.create_board()

		cinfo_path = rospy.get_param('~cinfo_path', '~/catkin_ws/src/clever/clever/camera_info/fisheye_cam_320.yaml')
		with open(cinfo_path) as cinfo:
			load = yaml.load(cinfo)
			self.camera_matrix = load.get('camera_matrix')
			self.correct_cm()

			self.dist_coeffs = load.get('distortion_coefficients')
			self.dist_coeffs = numpy.asarray(self.dist_coeffs['data'], dtype=numpy.float64)

		self.markers_detected = 0
		self.rvec = None
		self.tvec = None
		
		self.pose = None
		self.orient = None
		self.fail_cnt = None

		self.raw_img_sub = rospy.Subscriber('/main_camera/image_raw/throttled', Image, self.detect)

		self.img_pub = rospy.Publisher('~/RPS/debug', Image, queue_size=1)
		self.board_img_pub = rospy.Publisher('~/RPS/board_image', Image, queue_size=1)

		self.pose_pub = rospy.Publisher('~/RPS/pose', PoseStamped, queue_size=1) 

		rospy.loginfo('rps_pose inited')

#--- actions ---#

rospy.init_node('rps_pose')
rps = rps_class()

while not rospy.is_shutdown():
	rospy.sleep(0.1)

File: RPS/source/rps_pose.cpp
#include <ros/ros.h>
#include <stdio.h>
#include <leds_pose.h>
#include <opencv2/core.hpp>
#include <opencv2/opencv.hpp>
#include <opencv2/aruco.hpp>
#include <opencv2/aruco/dictionary.hpp>
#include <opencv2/calib3d/calib3d.hpp>
#include <opencv2/imgproc/imgproc.hpp>
#include <image_transport/image_transport.h>
#include <cv_bridge/cv_bridge.h>

#include <geometry_msgs/PoseStamped.h>
#include <sensor_msgs/Image.h>
#include <iostream>

using std::string;
using std::vector;

//---- class description ----//

class rpsPoseClass {

private:
	cv::Ptr<cv::aruco::Dictionary> dict;
    cv::Ptr<cv::aruco::DetectorParameters> param;
	cv::Ptr<cv::aruco::Board> board;
	cv::Vec3d pose, orient;
	int markers_detected;
	int fail_cnt, max_fail_cnt;
	int use_led_marker;
	int number_x, number_y, first_marker;
	float side, dist_x, dist_y;

	image_transport::CameraSubscriber img_sub;
	image_transport::Publisher img_pub;
	ros::Publisher board_pub;
	ros::Publisher pose_pub;

	void createBoard();
	void publishPoseMsg(int); 
	void parseCameraInfo(const sensor_msgs::CameraInfoConstPtr&, cv::Mat&, cv::Mat&);
	void detect(const sensor_msgs::ImageConstPtr&, const sensor_msgs::CameraInfoConstPtr&);

public:
	rpsPoseClass(int argc, char **argv) {
		ros::init(argc, argv, "rps_pose");
		ros::NodeHandle nh;
		ROS_INFO("initializing rps_pose...");

		nh.param("rps_pose/use_led_marker", use_led_marker, 0);
		nh.param("rps_pose/number_x", number_x, 1);
		nh.param("rps_pose/number_y", number_y, 1);
		nh.param("rps_pose/first_marker", first_marker, 0);
		nh.param<float>("rps_pose/side", side, 0.08);
		nh.param<float>("rps_pose/dist_x", dist_x, 0);
		nh.param<float>("rps_pose/dist_y", dist_y, 0);
		if (!use_led_marker) { 
			dict = cv::aruco::getPredefinedDictionary(cv::aruco::DICT_4X4_1000);
			param = cv::aruco::DetectorParameters::create();
			board = cv::makePtr<cv::aruco::Board>();
			board_pub = nh.advertise<sensor_msgs::Image>("/RPS/board_image", 1, true);
			createBoard();
		}

		fail_cnt = -1;
		nh.param("rps_pose/max_fail_cnt", max_fail_cnt, 5);

		image_transport::ImageTransport it(nh); 
		img_sub = it.subscribeCamera("/main_camera/throttled", 1, &rpsPoseClass::detect, this);
		
		img_pub = it.advertise("/RPS/debug", 1);
		pose_pub = nh.advertise<geometry_msgs::PoseStamped>("/RPS/pose", 1);
		ROS_INFO("rps_pose inited");
	}
};

//---- methods ----//

void rpsPoseClass::createBoard() {
	float max_y = side + (number_y - 1)*(side + dist_y);
	std::vector<int> markers_ids;
	markers_ids.resize(number_x*number_y);
	for (int i = 0; i < number_x*number_y; i++) {
		markers_ids.at(i) =  first_marker + i;
	}
	board->ids = markers_ids;

	for (int i = 0; i < number_y; i++ ) {
		for (int j = 0; j < number_x; j++) {
			std::vector<cv::Point3f> corners;
			corners.resize(4);
			corners[0] = cv::Point3f(j*(side + dist_x), max_y - i*(side + dist_y), 0);
			corners[1] = cv::Point3f(j*(side + dist_x) + side, max_y - i*(side + dist_y), 0);
			corners[2] = cv::Point3f(j*(side + dist_x) + side, max_y - i*(side + dist_y) - side, 0);
			corners[3] = cv::Point3f(j*(side + dist_x), max_y - i*(side + dist_y) - side, 0);
			board->objPoints.push_back(corners);
		} 
	}
	board->dictionary = dict;

	// debug
	cv::Mat board_image;
	cv_bridge::CvImage board_msg;
	cv::aruco::drawPlanarBoard(board, cv::Size(2000, 2000), board_image, 50, 1);
	cv::cvtColor(board_image, board_image, CV_GRAY2BGR);

	board_msg.image = board_image;
	board_msg.encoding = sensor_msgs::image_encodings::BGR8;
	board_pub.publish(board_msg.toImageMsg());

}

void rpsPoseClass::detect(const sensor_msgs::ImageConstPtr& msg, const sensor_msgs::CameraInfoConstPtr& cinfo) {
	
	// detect
	cv::Mat image = cv_bridge::toCvShare(msg, "bgr8")->image;
	cv::Mat camera_matrix(3, 3, CV_64F);
	cv::Mat dist_coeffs(8, 1, CV_64F);
	cv_bridge::CvImage img_msg;
	std::vector<std::vector<cv::Point2f>> corners;
	std::vector<int> ids;
	cv::Vec3d rvec, tvec;

	parseCameraInfo(cinfo, camera_matrix, dist_coeffs);
	if (!use_led_marker) {

		cv::aruco::detectMarkers(image, dict, corners, ids, param);

		if (ids.size() > 0) {
			markers_detected = cv::aruco::estimatePoseBoard(corners, ids, board, camera_matrix, dist_coeffs, rvec, tvec);
		} 
		else {
			markers_detected = 0;
		}

	} else {
		std::vector <cv::Point2f> led_corners;
		std::vector <cv::Vec3d> rvecs, tvecs;
		ledsPose::detectLedMarker(image, led_corners);

		if (led_corners.size() > 0) {
			corners.push_back(led_corners);
			ids.push_back(-1);
			cv::aruco::estimatePoseSingleMarkers(corners, side, camera_matrix, dist_coeffs, rvecs, tvecs);
			if (rvecs.size() > 0) {
				rvec = rvecs[0];
				tvec = tvecs[0];
				markers_detected = 1;
			} else {
				markers_detected = 0;
			}
		} else {
			markers_detected = 0;
		}
	}

	// decide if publish or not
	int valid;
	if (markers_detected > 0) {
		fail_cnt = 0;
		pose = -tvec;
		orient = rvec;
		valid = 1;
	} else {
		if (fail_cnt != -1) {
			valid = 1;
			fail_cnt++;
			if (fail_cnt > max_fail_cnt) {
				fail_cnt = -1;
				valid = 0;
			}
		} else {
			valid = 0;
		}
	}
	publishPoseMsg(valid);

	// debug
	if (img_pub.getNumSubscribers() > 0) {
		cv::aruco::drawDetectedMarkers(image, corners, ids);	
		if (markers_detected > 0) {
			cv::aruco::drawAxis(image, camera_matrix, dist_coeffs, rvec, tvec, 0.3);
		}
		img_msg.header.frame_id = msg->header.frame_id;
		img_msg.header.stamp = ros::Time::now();
		img_msg.encoding = sensor_msgs::image_encodings::BGR8;
		img_msg.image = image;
		img_pub.publish(img_msg.toImageMsg());
	}
}

void rpsPoseClass::parseCameraInfo(const sensor_msgs::CameraInfoConstPtr &cinfo, cv::Mat &camera_matrix, cv::Mat &dist_coeffs) {
    for (int i = 0; i < 3; i++ ) {
        for (int j = 0; j < 3; j++ ) {
            camera_matrix.at<double>(i, j) = cinfo->K[3 * i + j];
        }
    }
    for (int k = 0; k < cinfo->D.size(); k++ ) {
        dist_coeffs.at<double>(k) = cinfo->D[k];
    }
}

//void rpsPoseClass::calculateAndPublish() {
//}

void rpsPoseClass::publishPoseMsg(int valid) {
	geometry_msgs::PoseStamped pose_msg;
	pose_msg.header.stamp = ros::Time::now();
	if (valid) {
		pose_msg.header.frame_id = "main_camera";
		pose_msg.pose.position.x = pose[0];
		pose_msg.pose.position.y = pose[1];
		pose_msg.pose.position.z = pose[2];
		pose_msg.pose.orientation.x = orient[0];
		pose_msg.pose.orientation.y = orient[1];
		pose_msg.pose.orientation.z = orient[2];

	} else {
		pose_msg.header.frame_id = "";
	}

	pose_pub.publish(pose_msg);
}

//---- main ----//
int main(int argc, char **argv) {
	rpsPoseClass rps_pose(argc, argv);
	ros::spin();
	return 0;
}